Disallow column writes for delegated methods

saturnflyer opened the pull request:
When using `delegate_to_property_set` it sets the delegated property as an attribute of the main model so that errors may be tracked. Unfortunately, ActiveRecord will uses those attributes when writing data to the database from fixtures, for example. 

Prior to this PR, the code would attempt to write with `update_columns` using all of the model's attributes. These changes remove any attributes specified by `delegate_to_property_set` because they are not actual columns for the model.

In versions less than Rails 5.2, ActiveRecord will complain if an empty set of columns is provided by raising an ArgumentError. In Rails 5.2 and above, an empty set is allowed but an error will occur calling `first` and operating on it from the empty set.

srvance commented on lib/property_sets/active_record_extension.rb:
```diff
@@ -175,6 +175,12 @@ def update!(attributes)
 
+      def update_columns(attributes)
+        attrs = attributes.reject{|k,_| self.class.delegated_property_set_attributes.include?(k) }
+        raise ArgumentError, "Empty list of attributes to change" if attrs.empty?
```
Is it really an error if an update doesn't change the property set attributes?

saturnflyer replied:
I'm glad you asked... Rails 5.2 and above let the empty set go through but eventually hit an error. Below 5.2 and they handle that ArgumentError. So I just copied the behavior.
An alternative approach would be to have 2 different styles of testing for the different rails/activerecord versions.

saturnflyer commented:
Being verified in https://github.com/zendesk/zendesk/pull/43138

srvance commented on spec/support/account.rb:
```diff
@@ -47,3 +47,11 @@ class Account < ActiveRecord::Base
+
+module Other
+  class Account < ::Account
+    def self.delegated_property_set_attributes
```
So will you have to do this everywhere?

saturnflyer replied:
This is a copy of something we have to do in the main app. I went down the path of automatically including this behavior by default when using `delegate_to_property_set` but I don' think that we want to enforce that on every user of it.

The solution included making this a `class_attribute` which would be an inherited attribute for classes... but that seemed outside the scope of this module.

srvance replied:
Got it. I saw it in Classic and here and feared it would be necessary in all situations.

srvance commented on lib/property_sets/delegator.rb:
```diff
@@ -51,6 +51,8 @@ def delegate_to_property_set(setname, mappings)
           end
         end
+
+        puts self.user_provided_columns.reject!{|k,_| @delegated_property_set_attributes.include?(k.to_s) }
```
Did you mean to leave a `puts` here?

srvance commented:
Nice work!

grosser commented on lib/property_sets/active_record_extension.rb:
```diff
@@ -184,6 +184,38 @@ def update_property_set_attributes(attributes)
+          attributes
+        end
+
+        raise ArgumentError, "Empty list of attributes to change" if attrs.empty?
```
if `super` does not do this, then we should not do it either ?

saturnflyer replied:
It does in certain versions of rails. In later versions, rails just lets it blow up. Do we have a preference?

grosser replied:
do what rails does 🤷 
... not property_set gem responsibility to unify rails apis

grosser commented on lib/property_sets/active_record_extension.rb:
```diff
@@ -184,6 +184,38 @@ def update_property_set_attributes(attributes)
+
+      def filter_delegated_property_set_attributes(attribute_names)
+        if delegated_property_sets?
+          return attribute_names.map(&:to_s) - self.class.delegated_property_set_attributes.to_a
```
this is changing the input from symbols to strings, the update_columns was better in this regard since it only rejected

saturnflyer replied:
The `.map(&:to_s)` was only a defensive coercion in case future rails changed behavior. But I can pull that out.

grosser replied:
yeah either way is meh ... it's slow (using reject) or converts input (using to_s) or might fail later when not converting ... so 🤷 pick whatever you like :D

grosser commented on lib/property_sets/active_record_extension.rb:
```diff
@@ -184,6 +184,38 @@ def update_property_set_attributes(attributes)
       end
+
+      def update_columns(attributes)
+        attrs = if delegated_property_sets?
+          attributes.reject{|k,_| self.class.delegated_property_set_attributes.include?(k.to_s) }
+        else
+          attributes
+        end
```
```suggestion
        if delegated_property_sets?
          attributes = attributes.reject{|k,_| self.class.delegated_property_set_attributes.include?(k.to_s) }
        end
```
... then use `attributes` below

grosser commented:
ideally get rid of delegation since it's causing issues ... less layers of indirection would be better 😞

grosser commented:
other alternative: stop declaring things as attributes ? ... maybe be can make only the errors bug go away so we can declare errors on non-attributes

saturnflyer commented:
@grosser I addressed some of your feedback. Changing the way we use this might be worth an exploration but my goal with this PR is to just get in functioning and get past errors that were revealed by the last release.
